A14-channelz.md

```diff
@@ -0,0 +1,1051 @@
+Note that only the connection level flow control window is reported, rather than
+the stream level.  TCP level flow and congestion control counters are exposed
+via socket options.
+
+#### Socket Options
+
+At this low level in the gRPC stack, the implementation difference begin vary
+more greatly.  Each OS or platform may have differing amounts of socket
+information available.  To address these differences, each socket option name
+and value is expressed as a string.
+
+If there is a more structured form (such as TCP info), it too can be included
+in the socket option message.  At least one of the string form or the structured
+form should be present.  The additional data is exposed as a Proto3 Any field,
+to accommodate types added later.  A few default messages are included for the
+common options, such as that for TCP Info, and socket timeouts.
+
+When querying for socket option data, as much detail should be included as
+possible.  Since there are many types of socket options (at least 50),  the
+number of options included in the socket data may be large.  If a querying a
+socket option returns an error, it should not be fatal.  The channelz info is a
+debugging tool, so it should try to ignore the failed socket option and continue
+on.
+
+This means that varying amounts of socket data may be returned by the channelz
+service.  Implementations may avoid querying the options if it would be
+impractical, though they are encouraged to do so.
+
+### Addresses
+
+An important part of debugging a connection is finding "who" the socket is
+connected to.  Toanswer this question each socket exposes a remote and a local
+address field.  The address implies what kind of socket it is (e.g. TCP, UDP,
+UDS, etc.) as well as the the identity used to establish the connection (e.g.
+IP Address, port number, filename, etc.).
+
+Not every socket has a remote address;  listening sockets may only include a
+local address.  Not every socket has a local address; UDS sockets may only
+include a remote address.
+
+By default, only TCP and UDS addresses are defined.  Because there are many
+kinds of address families that cannot be anticipated, the address includes an
+extendable proto Any field for extension.
+
+### Security
+
+Cryptographically secure connections are expected for use with gRPC.  The
+details of the crypto used on the connection are generally useful, and are thus
+exposed on the Socket message.  As with Socket options and Addresses, there may
+be many kinds of security that exist outside of the commonly used TLS, so the
+structure is extensible.  By default, only a TLS message is defined.
+
+# Querying Channelz
+
+Channelz exposes four methods to access channelz data:
+
+* GetTopChannels returns the root channels for channel trees
+* GetServers returns the servers in the system
+* GetChannel returns info about a specific channel
+* GetSubchannel returns info about a specific subchannel
```

Comment:
I think that we need some way to paginate data about subchannels.  For example, in the common case of a client with a single top-level channel and a large number of subchannels, paginating the subchannels is actually more important than paginating the top-level channels.

Reply:
Why do you think it's unlikely?  Do you doubt the existence of clients that talk to large numbers of backends?  I'd be very surprised if we have data showing that this is not a common case.

Reply:
more than 100 subchannels for a single channel.   I do doubt that high subchannel usages exist.    We should optimize for the common case, not for the rare one.  This means making users who have excess subchannels pay a slightly higher cost when they query the channelz service.

Also IIRC, the default number of addresses returned by the internal load balancer was 20.